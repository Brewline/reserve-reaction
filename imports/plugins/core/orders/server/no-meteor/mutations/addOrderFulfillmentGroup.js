import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import { Order as OrderSchema } from "/imports/collections/schemas";
import { orderFulfillmentGroupInputSchema, orderItemInputSchema } from "../simpleSchemas";
import buildOrderFulfillmentGroupFromInput from "../util/buildOrderFulfillmentGroupFromInput";
import updateGroupStatusFromItemStatus from "../util/updateGroupStatusFromItemStatus";
import updateGroupTotals from "../util/updateGroupTotals";

const groupInputSchema = orderFulfillmentGroupInputSchema.clone().extend({
  // Make items optional since we have the option of moving items
  // from another group
  "items": {
    type: Array,
    optional: true,
    minCount: 1
  },
  "items.$": orderItemInputSchema
});

const inputSchema = new SimpleSchema({
  "fulfillmentGroup": groupInputSchema,
  "moveItemIds": {
    type: Array,
    optional: true,
    minCount: 1
  },
  "moveItemIds.$": String,
  "orderId": String
});

/**
 * @method addOrderFulfillmentGroup
 * @summary Use this mutation to add a new order fulfillment group to an order. It must have at least one
 *   item, which can be provided or moved from another existing group.
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Necessary input. See SimpleSchema
 * @return {Promise<Object>} Object with `order` property containing the updated order and a
 *   `newFulfillmentGroupId` property set to the ID of the added group
 */
export default async function addOrderFulfillmentGroup(context, input) {
  inputSchema.validate(input);

  const {
    fulfillmentGroup: inputGroup,
    moveItemIds,
    orderId
  } = input;

  const { appEvents, collections, isInternalCall, userHasPermission, userId } = context;
  const { Orders } = collections;

  // First verify that this order actually exists
  const order = await Orders.findOne({ _id: orderId });
  if (!order) throw new ReactionError("not-found", "Order not found");

  // Allow update if the account has "orders" permission. When called internally by another
  // plugin, context.isInternalCall can be set to `true` to disable this check.
  if (!isInternalCall && !userHasPermission(["orders"], order.shopId)) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  const { billingAddress, cartId, currencyCode } = order;

  // If there are moveItemIds, find and pull them from their current groups
  let updatedGroups;
  const orderSurcharges = [];
  const movingItems = [];
  if (moveItemIds) {
    updatedGroups = await Promise.all(order.shipping.map(async (group) => {
      let movedSomeItems = false;
      const updatedItems = group.items.reduce((list, item) => {
        if (moveItemIds.includes(item._id)) {
          movingItems.push(item);
          movedSomeItems = true;
        } else {
          list.push(item);
        }
        return list;
      }, []);

      if (!movedSomeItems) return group;

      if (updatedItems.length === 0) {
        throw new ReactionError("invalid-param", "moveItemIds would result in group having no items");
      }

      const updatedGroup = {
        ...group,
        // There is a convenience itemIds prop, so update that, too
        itemIds: updatedItems.map((item) => item._id),
        items: updatedItems,
        totalItemQuantity: updatedItems.reduce((sum, item) => sum + item.quantity, 0)
      };

      // Update group shipping, tax, totals, etc.
      const { groupSurcharges } = await updateGroupTotals(context, {
        billingAddress,
        cartId,
        currencyCode,
        discountTotal: updatedGroup.invoice.discounts,
        group: updatedGroup,
        orderId,
        selectedFulfillmentMethodId: updatedGroup.shipmentMethod._id
      });

      // Push all group surcharges to overall order surcharge array.
      // Currently, we do not save surcharges per group
      orderSurcharges.push(...groupSurcharges);

      // Ensure proper group status
      updateGroupStatusFromItemStatus(updatedGroup);

      // Return the group, with items and workflow potentially updated.
      return updatedGroup;
    }));

    if (moveItemIds.length !== movingItems.length) {
      throw new ReactionError("invalid-param", "Some moveItemIds did not match any item IDs on the order");
    }
  } else {
    updatedGroups = [...order.shipping];
    if (Array.isArray(order.surcharges)) orderSurcharges.push(...order.surcharges);
  }

  // Now build the new group we are adding
  const { group: newGroup, groupSurcharges } = await buildOrderFulfillmentGroupFromInput(context, {
    // If we are moving any items from existing groups to this new group, push those into
    // the newGroup items array.
    additionalItems: movingItems,
    billingAddress,
    cartId,
    currencyCode,
    // No support for discounts for now. Pending future promotions revamp.
    discountTotal: 0,
    inputGroup,
    orderId
  });

  // Add the new group to the order groups list
  updatedGroups.push(newGroup);

  // Push all group surcharges to overall order surcharge array.
  // Currently, we do not save surcharges per group
  orderSurcharges.push(...groupSurcharges);

  // Now we're ready to update the database
  const modifier = {
    $set: {
      shipping: updatedGroups,
      surcharges: orderSurcharges,
      totalItemQuantity: updatedGroups.reduce((sum, group) => sum + group.totalItemQuantity, 0),
      updatedAt: new Date()
    }
  };

  OrderSchema.validate(modifier, { modifier: true });

  const { modifiedCount, value: updatedOrder } = await Orders.findOneAndUpdate(
    { _id: orderId },
    modifier,
    { returnOriginal: false }
  );
  if (modifiedCount === 0 || !updatedOrder) throw new ReactionError("server-error", "Unable to update order");

  await appEvents.emit("afterOrderUpdate", {
    order: updatedOrder,
    updatedBy: userId
  });

  return { newFulfillmentGroupId: newGroup._id, order: updatedOrder };
}
