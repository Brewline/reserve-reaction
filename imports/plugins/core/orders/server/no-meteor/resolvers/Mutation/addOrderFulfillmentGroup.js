import { decodeCartItemsOpaqueIds } from "@reactioncommerce/reaction-graphql-xforms/cart";
import { decodeFulfillmentMethodOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/fulfillment";
import { decodeOrderOpaqueId, decodeOrderItemOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/order";
import { decodeShopOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/shop";

/**
 * @name Mutation/addOrderFulfillmentGroup
 * @method
 * @memberof Payments/GraphQL
 * @summary resolver for the addOrderFulfillmentGroup GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {Object} args.input.fulfillmentGroup The order fulfillment group input, used to build the new group
 * @param {String[]} [args.input.moveItemIds] Optional list of order item IDs that should be moved from an
 *   existing group to the new group.
 * @param {String} args.input.orderId Order ID
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} AddOrderFulfillmentGroupPayload
 */
export default async function addOrderFulfillmentGroup(parentResult, { input }, context) {
  const {
    clientMutationId = null,
    fulfillmentGroup,
    moveItemIds,
    orderId
  } = input;

  const { newFulfillmentGroupId, order } = await context.mutations.addOrderFulfillmentGroup(context, {
    fulfillmentGroup: {
      ...fulfillmentGroup,
      items: fulfillmentGroup.items ? decodeCartItemsOpaqueIds(fulfillmentGroup.items) : null,
      selectedFulfillmentMethodId: decodeFulfillmentMethodOpaqueId(fulfillmentGroup.selectedFulfillmentMethodId),
      shopId: decodeShopOpaqueId(fulfillmentGroup.shopId)
    },
    moveItemIds: moveItemIds && moveItemIds.map(decodeOrderItemOpaqueId),
    orderId: decodeOrderOpaqueId(orderId)
  });

  return {
    clientMutationId,
    newFulfillmentGroupId,
    order
  };
}
