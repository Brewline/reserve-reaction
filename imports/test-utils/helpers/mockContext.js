const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
