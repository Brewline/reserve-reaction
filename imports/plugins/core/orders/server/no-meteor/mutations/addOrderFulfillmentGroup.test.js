import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import {
  restore as restore$buildOrderFulfillmentGroupFromInput,
  rewire as rewire$buildOrderFulfillmentGroupFromInput
} from "../util/buildOrderFulfillmentGroupFromInput";
import {
  restore as restore$updateGroupTotals,
  rewire as rewire$updateGroupTotals
} from "../util/updateGroupTotals";
import addOrderFulfillmentGroup from "./addOrderFulfillmentGroup";

beforeEach(() => {
  jest.resetAllMocks();
});

afterEach(() => {
  restore$buildOrderFulfillmentGroupFromInput();
  restore$updateGroupTotals();
});

test("throws if orderId isn't supplied", async () => {
  const fulfillmentGroup = Factory.orderFulfillmentGroupInputSchema.makeOne({});

  // There is a bug where Factory always adds _id
  delete fulfillmentGroup._id;

  await expect(addOrderFulfillmentGroup(mockContext, { fulfillmentGroup })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if fulfillmentGroup isn't supplied", async () => {
  await expect(addOrderFulfillmentGroup(mockContext, { orderId: "123" })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(null));

  const fulfillmentGroup = Factory.orderFulfillmentGroupInputSchema.makeOne({});

  // There is a bug where Factory always adds _id
  delete fulfillmentGroup._id;

  await expect(addOrderFulfillmentGroup(mockContext, {
    fulfillmentGroup,
    orderId: "order1"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if permission check fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        _id: "123",
        items: []
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(false);

  const fulfillmentGroup = Factory.orderFulfillmentGroupInputSchema.makeOne({});

  // There is a bug where Factory always adds _id
  delete fulfillmentGroup._id;

  await expect(addOrderFulfillmentGroup(mockContext, {
    fulfillmentGroup,
    orderId: "order1"
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders"], "SHOP_ID");
});

test("throws if an item ID being moved does not exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "123",
        items: Factory.OrderItem.makeMany(2, {
          _id: (index) => `ITEM_1_${index}`
        })
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "345",
        items: Factory.OrderItem.makeMany(2, {
          _id: (index) => `ITEM_2_${index}`
        })
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  const mockUpdateGroupTotals = jest.fn().mockName("updateGroupTotals").mockReturnValue(Promise.resolve({ groupSurcharges: [] }));
  rewire$updateGroupTotals(mockUpdateGroupTotals);

  const fulfillmentGroup = Factory.orderFulfillmentGroupInputSchema.makeOne({});

  // There is a bug where Factory always adds _id
  delete fulfillmentGroup._id;

  await expect(addOrderFulfillmentGroup(mockContext, {
    fulfillmentGroup,
    moveItemIds: ["xyz"],
    orderId: "order1"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("skips permission check if context.isInternalCall", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        _id: "123",
        items: Factory.OrderItem.makeMany(2, {
          workflow: {
            status: "new",
            workflow: ["new"]
          }
        })
      }),
      Factory.OrderFulfillmentGroup.makeOne({
        items: Factory.OrderItem.makeMany(2, {
          workflow: {
            status: "new",
            workflow: ["new"]
          }
        })
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  mockContext.isInternalCall = true;

  const mockBuildOrderFulfillmentGroupFromInput = jest.fn().mockName("buildOrderFulfillmentGroupFromInput");
  mockBuildOrderFulfillmentGroupFromInput.mockReturnValueOnce(Promise.resolve({
    group: Factory.OrderFulfillmentGroup.makeOne({
      items: [
        Factory.OrderItem.makeOne({
          _id: "ITEM_1",
          quantity: 1,
          price: {
            amount: 1,
            currencyCode: "USD"
          },
          subtotal: 1,
          workflow: {
            status: "new",
            workflow: ["new"]
          }
        })
      ],
      workflow: {
        status: "new",
        workflow: ["new"]
      }
    }),
    groupSurcharges: []
  }));
  rewire$buildOrderFulfillmentGroupFromInput(mockBuildOrderFulfillmentGroupFromInput);

  const fulfillmentGroup = Factory.orderFulfillmentGroupInputSchema.makeOne({});

  // There is a bug where Factory always adds _id
  delete fulfillmentGroup._id;

  await addOrderFulfillmentGroup(mockContext, {
    fulfillmentGroup,
    orderId: "order1"
  });

  delete mockContext.isInternalCall;

  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
});

test("adds an order fulfillment group", async () => {
  const originalGroups = [
    Factory.OrderFulfillmentGroup.makeOne({
      _id: "123",
      items: Factory.OrderItem.makeMany(2, {
        quantity: 1,
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }),
      totalItemQuantity: 2
    }),
    Factory.OrderFulfillmentGroup.makeOne({
      items: Factory.OrderItem.makeMany(2, {
        _id: "345",
        quantity: 1,
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }),
      totalItemQuantity: 2
    })
  ];

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: originalGroups,
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  const newGroup = Factory.OrderFulfillmentGroup.makeOne({
    items: [
      Factory.OrderItem.makeOne({
        _id: "ITEM_1",
        quantity: 1,
        price: {
          amount: 1,
          currencyCode: "USD"
        },
        subtotal: 1,
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    totalItemQuantity: 1,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });
  const mockBuildOrderFulfillmentGroupFromInput = jest.fn().mockName("buildOrderFulfillmentGroupFromInput");
  mockBuildOrderFulfillmentGroupFromInput.mockReturnValueOnce(Promise.resolve({
    group: newGroup,
    groupSurcharges: []
  }));
  rewire$buildOrderFulfillmentGroupFromInput(mockBuildOrderFulfillmentGroupFromInput);

  const fulfillmentGroup = Factory.orderFulfillmentGroupInputSchema.makeOne({});

  // There is a bug where Factory always adds _id
  delete fulfillmentGroup._id;

  await addOrderFulfillmentGroup(mockContext, {
    fulfillmentGroup,
    orderId: "order1"
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "order1" },
    {
      $set: {
        shipping: [
          ...originalGroups,
          newGroup
        ],
        surcharges: [],
        totalItemQuantity: 5,
        updatedAt: jasmine.any(Date)
      }
    },
    { returnOriginal: false }
  );
});
