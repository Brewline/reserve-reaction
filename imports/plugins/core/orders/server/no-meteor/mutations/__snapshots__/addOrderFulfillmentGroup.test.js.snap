// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if an item ID being moved does not exist 1`] = `"Some moveItemIds did not match any item IDs on the order"`;

exports[`throws if fulfillmentGroup isn't supplied 1`] = `"Fulfillment group is required"`;

exports[`throws if orderId isn't supplied 1`] = `"Order ID is required"`;

exports[`throws if permission check fails 1`] = `"Access Denied"`;

exports[`throws if the order doesn't exist 1`] = `"Order not found"`;
